driftbase/flexmatch.py

```diff
@@ -0,0 +1,558 @@
+                    continue
+                player_ticket["Status"] = "TIMED_OUT"
+                ticket_lock.ticket = player_ticket
+                players_to_notify.add(player_id)
+    _post_matchmaking_event_to_members(players_to_notify, "MatchmakingFailed", event_data={"reason": "TimeOut"})
+
+def _process_matchmaking_failed_event(event):
+    # FIXME: This is pretty much the same as a timeout; refactor
+    players_to_notify = set()
+    for ticket in event["tickets"]:
+        ticket_id = ticket["ticketId"]
+        for player in ticket["players"]:
+            player_id = int(player["playerId"])
+            ticket_key = _get_player_ticket_key(player_id)
+            with _LockedTicket(ticket_key) as ticket_lock:
+                player_ticket = ticket_lock.ticket
+                if player_ticket is None:
+                    log.warning(f"Failed event for ticket {ticket_id} includes player {player_id} who has no ticket.")
+                    continue
+                if ticket_id != player_ticket["TicketId"]:
+                    # Maybe a failure on a backfill ticket?
+                    log.warning(f"Failed event for ticket {ticket_id} includes player {player_id} who has ticket {player_ticket['TicketId']}.")
+                    continue
+                if player_ticket.get("GameSessionConnectionInfo", None) is not None:
+                    # If we've recorded a session, then the player has been placed in a match already
+                    log.info(f"Player {player_id} has a session attached to his ticket.  Timeout is nonsensical here. Ignoring.")
+                    continue
+                player_ticket["Status"] = "FAILED"
+                ticket_lock.ticket = player_ticket
+                players_to_notify.add(player_id)
+    _post_matchmaking_event_to_members(players_to_notify, "MatchmakingFailed", event_data={"reason": event["reason"]})
+
+
+class GameLiftRegionClient(object):
+    __gamelift_clients_by_region = {}
+    __gamelift_sessions_by_region = {}
+
+    def __init__(self, region):
+        self.region = region
+        client = self.__class__.__gamelift_clients_by_region.get(region)
+        if client is None:
+            session = self.__class__.__gamelift_sessions_by_region.get(region)
+            if session is None:
+                session = boto3.Session(region_name=self.region)
+                role_to_assume = _get_gamelift_role()
+                if role_to_assume:
+                    session = assume_role(session, role_to_assume)
+                self.__class__.__gamelift_sessions_by_region[region] = session
+            client = session.client("gamelift")
+            self.__class__.__gamelift_clients_by_region[region] = client
+
+    def __getattr__(self, item):
+        return getattr(self.__class__.__gamelift_clients_by_region[self.region], item)
+
+
+class _LockedTicket(object):
```

Comment:
heh, probably not; I just didn't know about it when I wrote this; I googled redis key locking and somehow ended up on a StackOverflow answer saying there was no such thing and I'd have to roll my own :P  Been meaning to plug it in for my impl since I noticed it being there but just forgot to do it.  I'll get to it ;)